---
title: "Basic loops"
author: "Min-Yao"
date: "`r Sys.Date()`"
output: 
  html_document: 
    keep_md: yes
---

## Learning Goals
*At the end of this exercise, you will be able to:*  
1. know the concept of loops.
2. understand the three different types of loops.
3. use loops in vectors and data frames.
4. use loops through multiple files and create a summary.

## Resources and references
- [Statistics Globe](https://statisticsglobe.com/loops-in-r/)
- [R - Loops](https://www.tutorialspoint.com/r/r_loops.htm)
- [Data Carpentry for Biologists](https://datacarpentry.org/semester-biology/materials/for-loops-R/)
- [Data Carpentry](https://github.com/datacarpentry/semester-biology)
- [Data Science in Omics Introduction](https://hoytpr.github.io/bioinformatics-semester/)

## Load the libraries
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(ggmap)
library(lubridate)
```

## What Are Loops?
A loop is a programming instruction that repeats a specific code chunk in a process until a specific condition is reached. Each time the code block within the loop is executed is called an **iteration.**

*Different Types of Loops:*
Depending on your specific programming situation, you may need different loop-structures that execute the code blocks within the loop on the basis of different conditions. The R programming language generally provides three different types of loops: 
1. Repeat-loops, 
2. While-loops, 
3. For-loops

## Repeat-Loops
Repeat-loops repeat a code block until a break condition is fulfilled. This break condition marks the end of the loop.
![Repeat-Loops](repeat-loop-in-r.png)

### Syntax of repeat loop
```{r,eval=FALSE}
repeat { 
   commands # this is the code you want the loop to run
   if(condition) { # this is the break conditions
      break
   }
}
```

### example 0: Writing a repeat-Loop in R (Basics)
```{r}
x <- 0                 # Preliminary specification of data object
repeat{                # Head of repeat-loop
  x <- x + 1           # Body of repeat-loop
  print(x)
  if(x >= 10) {        # Break condition of repeat-loop
    break
  }
}
```
This printed out 1-10!

Starting from 0, when you add 1, it is now x = 1. R will check the break condition: 1 is not larger than 10, so the loop is executed again. Once it hit 10, this did meet the break condition, so the loop was exited.

### example 1: repeat-Loop Through Columns of Data Frame
```{r}
data <- data.frame(x1 = 1:5,      # Create example data
                   x2 = 6:10,
                   x3 = 11:15)
data # we will use a repeat loop to change the values

data1 <- data                   # Replicate example data
i <- 1                          # Create running index (is like 'x' in the previous example)
repeat{                         # Start repeat-loop
  data1[ , i] <- data1[ , i] + 100 # the value inside the data frame structure based on the column... we are taking the Ith column, it's value, and adding 100 to that value. it is then stored in the original column to change to value
  i <- i + 1 # i will add 1 to the index
  if(i > 2) {
    break
  }
}
data1
```
The value is 100 more than the original value.

### Practice: Please write a repeat loop to change all numeric data from in centimeters to millimeters.
```{r}
#?iris
data(iris)                          # Loading data frame
head(iris)                          # Head of example data
```

My work
```{r}
iris1 <- iris # Replicate example data
i <- 1 # step 1: create running index with 1
repeat{ # begin repeat statement
  iris1[ , i] <- iris1[ , i] * 10 # code we want to run
  i <- i + 1
  if(i > 4) { # change the index value for each time the code is run
    break
  }
}
iris1
```


Min-Yao work
```{r}
iris_my <- iris
i <- 1
repeat{
  iris_my[ , i] <- iris_my[ , i] * 10
  i <- i + 1
  if (!is.numeric(iris_my[ , i])) {
    break
  }
}
head(iris_my)
```


## While-Loops
While-loops repeat a code block as long as a certain logical condition is TRUE. It tests the condition before executing the loop body.
![While-Loops](while-loop-in-r.png)

The code will run until the condition is turned to `FALSE`.

### Syntax of while loop
```{r,eval=FALSE}
while (test_expression) { # while something is true... (the logical condition)
   statement #do this code block
}
```

### example 0: Writing while-Loop in R (Basics)
```{r}
x <- 0                  # Preliminary specification of data object
while(x < 10) {         # Head of while-loop: logical condition is that x < 10
  x <- x + 1            # Body of while-loop: add 1 to the x value and save this total value into x, then print the new value
  print(x)
}
```

### example 1: while-Loop Through Columns of Data Frame
```{r}
data2 <- data                       # Replicate example data
data2
i <- 1                              # Create running index
while(i <= 2) {                     # Start while-loop: condition is that i is less than or equal to 2
  data2[ , i] <- data2[ , i] + 100 # our code says that we take the specific column and add 100 in this value, then save it
  i <- i + 1
}
data2
```

### Practice: Please write a while loop to change all numeric data from in centimeters to millimeters.
```{r}
head(iris)                                        # Showing head of data in RStudio console
iris2 <- iris                                     # Replicate example data
```

```{r}
iris2
i <- 1
while (is.numeric(iris2[, i])) { # run the loop when the specific column has numeric values
  iris2[ , i] <- iris2[ , i] * 10
  i <- i + 1
}
head(iris2)
```


## For-Loops
For-loops specify a collection of objects (e.g. elements in a vector or list) to which a code block should be applied.
![For-Loops](for-loop-in-r.png)

At the beginning, the initialization is all the objects you will be working with. After this, there is a code block (body of the loop). After this, the for loop will check if the last object from the beginning. If this is true, it will stop. This is not just limited to integers or numbers -- we can also pass character vectors, logical vectors, lists, and expressions. "For" means "for each item in the collection, we are going to do something to this item."

### Syntax of for loop
```{r,eval=FALSE}
for (value in vector) { #inside the (), there is a value *in* a vector (or collection of objects)
   statements #the code block that we want to execute
}
```

### example 0: Loop Through Vector in R (Basics)
```{r}
for (i in 1:10) {
  print(i)
}
```

### example 1: For-Loop Through Columns of Data Frame
```{r}
data3 <- data                       # Replicate example data
data3
for(i in 1:2) {           # for-loop over columns
  data3[ , i] <- data3[ , i] + 100
}
data3
```

### Practice 1: Please write a for loop to change only *width data* from in centimeters to millimeters.
My work.. I only changed sepal width. Not completely correct!
```{r}                                                 
head(iris)                                             # Inspecting iris flower data set
iris3 <- iris                                          # Replicate iris data set
for(i in 2){
  iris3[ , i] <- iris3[ , i] * 10
}
iris3
```

Min-Yao's work
```{r}
for (i in c(2,4)) {
  iris3[ , i] <- iris3[ , i] * 10
}
iris3
```

```{r}
iris3 <- iris 
for (i in 1:ncol(iris3)) {
  if(str_detect(colnames(iris3[i]), "Width")){
  iris3[ , i] <- iris3[ , i] * 10  
  }
}
iris3
```


### example 2: Looping over files
The simulated satellite collar data is downloaded from [Data Carpentry](http://www.datacarpentry.org/semester-biology/data/locations.zip).

We can get the names of each of the files we want to loop over by using `list.files()`.
```{r}
data_files <- list.files("data/locations", pattern = "locations-2*", full.names = TRUE)
data_files
```

We would like to count the number of observations in each file.
```{r message=FALSE, warning=FALSE}

# create an *empty* vector to store those counts
results <- vector(mode = "integer", length = length(data_files)) # this will store integers in this vector, and the length will be as long as the data_files (what we created above)

for (i in 1:length(data_files)){ # we will loop from 1 to the length of the data_files vector
  data0 <- read_csv(data_files[i]) # we will read the .csv file in the locations we found
  count <- nrow(data0) # we will count how many rows are inside the data
  results[i] <- count # we will save the count into the results vector
}
results
```

### Practice 2: Storing loop results in a data frame
```{r}
# Start by creating an empty data frame
results <- data.frame(file_name = vector(mode = "character", length = length(data_files)), # first column is file+name, length is the second column, and the third column is the count
                      count = vector(mode = "integer", length = length(data_files)))
for (i in 1:length(data_files)) {
  data0 <- read_csv(data_files[i])
  count <- nrow(data0) # now we need to specify what we want in the dataframe
  results$file_name[i] <- data_files[i]
  results$count[i] <- count
}
results
```

### example 3: Looping over files for ggplot
Letâ€™s use some satellite collar data on a number of different individuals and we want to be able to quickly look at all of their recent movements at once.
```{r message=FALSE, warning=FALSE}
data_files3 <- list.files("data/individual_collar_data", full.names = TRUE)
data_files3

# only ggplot
for (i in 1:length(data_files3)){ # for each object in the vector data_files3 (1-10)
  data0 <- as.data.frame(read_csv(data_files3[i])) # you need to read the csv and then turn it into dataframe... then save it!
  print( # note it will not show the maps unless you say something inside the loop!
    ggplot(data0, aes(x=long,y=lat))+
      geom_path()+
      geom_point()+
      labs(title = data_files3[i], # the title is generated from the file_name
           x = "longitude", 
           y = "latitude")
  )
} 
```

How do we put them all in the map?

First, we need to create one giant dataframe with all of our data in it.
```{r}
data_list <- lapply(data_files3, read_csv)
collar_data_all <- bind_rows(data_list)
collar_data_all
```

Now, we need to find the maximum and minimum for latitude and longitude.
```{r}
collar_data_all %>% 
  summarise(max_lat = max(lat),
            min_lat = min(lat),
            max_long = max(long),
            min_long = min(long))
```

Define the boundary box.
```{r}
lat <- c(14.91539, 31.56526)
long <- c(122.0874, 134.9906)
bbox <- make_bbox(long, lat, f = 0.5)
```

Make the map! (REDO THIS)
```{r}
map <- get_map(bbox, maptype = "terrain", source = "stamen")
ggmap(map)
for (i in 1:length(data_files3)) {
  data <- as.data.frame(read_csv(data_files3[i]))
  print(
    ggmap(map) +
      geom_path(data = data0, aes(long, lat))+
      geom_point(data = data0, aes(long, lat))+
      labs(title = data_files3[i],
           x = "longitude",
           y = "latitude")
  )
}
```


### example 4: Looping over files for basic genetic analysis
[Bioconductor](https://www.bioconductor.org/) project is an open source software repository that hosts a wide range of statistical tools developed in the R programming environment. Please use standard R installation procedures to install the BiocManager package and then install "ShortRead" and "Biostrings". 
```{r message=FALSE, warning=FALSE}
 if (!requireNamespace("BiocManager", quietly = TRUE))
     install.packages("BiocManager")
 BiocManager::install()                            # To install core packages or update installed packages
 BiocManager::install(c("ShortRead", "Biostrings"))# Install specific packages
library(ShortRead)
library(Biostrings)
```

The data is downloaded from [PATRIC, BACTERIAL BIOINFORMATICS RESOURCE CENTER](https://patricbrc.org/)
The first two lines load a .[fasta](https://en.wikipedia.org/wiki/FASTA_format) file using the ShortRead package in Bioconductor. The second two lines determine the frequency of all of the bases in sequence and then calculate the GC content.
```{r}
reads <- readFasta("data/archaea-dna/a-saccharovorans.fasta")
seq <- sread(reads)
base_freq <- alphabetFrequency(seq)
gc_content <- (base_freq[1, "G"] + base_freq[1, "C"]) / sum(base_freq) * 100
gc_content
```

What if you have many fasta files? Don't do this... use a loop!
Then create a for loop that uses the above code to read in each sequence file and calculate itâ€™s GC content. Store the resulting values in a data frame with one column with file names and a second column with GC contents.
```{r}
data_files2 <- list.files("data/archaea-dna")
data_files2

GC_results <- data.frame(file_name = vector(mode = "character", length = length(data_files2)),
                         gc_content = vector(mode = "integer", length = length(data_files2)))

for(i in 1:length(data_files2)){
  reads <- readFasta("data/archaea-dna", data_files2[i])
  seq <- sread(reads)
  base_freq <- alphabetFrequency(seq)
  gc_content <- (base_freq[1, "G"] + base_freq[1, "C"]) / sum(base_freq) * 100
  GC_results$file_name[i]
}
```

## Packages for next time  
Please install packages below for part 2!  
```{r}
#install.packages("qtl")
#install.packages("qtlcharts")
```

## That's it! Take a break and I will see you on Zoom!  

-->[Home](https://jmledford3115.github.io/datascibiol/)